storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncSupervisorEvent.java

```diff
@@ -0,0 +1,626 @@
+
+        if (resourcesJar != null) {
+            LOG.info("Extracting resources from jar at {} to {}", resourcesJar, targetDir);
+            Utils.extractDirFromJar(resourcesJar, ConfigUtils.RESOURCES_SUBDIR, stormroot);
+        } else if (url != null) {
+
+            LOG.info("Copying resources at {} to {} ", url.toString(), targetDir);
+            if (url.getProtocol() == "jar") {
+                JarURLConnection urlConnection = (JarURLConnection) url.openConnection();
+                Utils.extractDirFromJar(urlConnection.getJarFileURL().getFile(), ConfigUtils.RESOURCES_SUBDIR, stormroot);
+            } else {
+                FileUtils.copyDirectory(new File(url.getFile()), (new File(targetDir)));
+            }
+        }
+    }
+
+    /**
+     * Downloading to permanent location is atomic
+     * 
+     * @param conf
+     * @param stormId
+     * @param masterCodeDir
+     * @param localizer
+     * @throws Exception
+     */
+    private void downloadDistributeStormCode(Map conf, String stormId, String masterCodeDir, Localizer localizer) throws Exception {
+
+        String tmproot = ConfigUtils.supervisorTmpDir(conf) + Utils.FILE_PATH_SEPARATOR + Utils.uuid();
+        String stormroot = ConfigUtils.supervisorStormDistRoot(conf, stormId);
+        ClientBlobStore blobStore = Utils.getClientBlobStoreForSupervisor(conf);
+        FileUtils.forceMkdir(new File(tmproot));
+        if (Utils.isOnWindows()) {
+            if (Utils.getBoolean(conf.get(Config.SUPERVISOR_RUN_WORKER_AS_USER), false)) {
+                throw new RuntimeException("ERROR: Windows doesn't implement setting the correct permissions");
+            }
+        } else {
+            Utils.restrictPermissions(tmproot);
+        }
+        String stormJarKey = ConfigUtils.masterStormJarKey(stormId);
+        String stormCodeKey = ConfigUtils.masterStormCodeKey(stormId);
+        String stormConfKey = ConfigUtils.masterStormConfKey(stormId);
+        String jarPath = ConfigUtils.supervisorStormJarPath(tmproot);
+        String codePath = ConfigUtils.supervisorStormCodePath(tmproot);
+        String confPath = ConfigUtils.supervisorStormConfPath(tmproot);
+        Utils.downloadResourcesAsSupervisor(stormJarKey, jarPath, blobStore);
+        Utils.downloadResourcesAsSupervisor(stormCodeKey, codePath, blobStore);
+        Utils.downloadResourcesAsSupervisor(stormConfKey, confPath, blobStore);
+        blobStore.shutdown();
+        Utils.extractDirFromJar(jarPath, ConfigUtils.RESOURCES_SUBDIR, tmproot);
+        downloadBlobsForTopology(conf, confPath, localizer, tmproot);
+        if (IsDownloadBlobsForTopologySucceed(confPath, tmproot)) {
+            LOG.info("Successfully downloaded blob resources for storm-id {}", stormId);
+            FileUtils.forceMkdir(new File(stormroot));
+            Files.move(new File(tmproot).toPath(), new File(stormroot).toPath(), StandardCopyOption.ATOMIC_MOVE);
```

Comment:
We are missing some windows specific changes here.

```
        (if (Utils/isOnWindows)
          ; Files/move with non-empty directory doesn't work well on Windows
          (FileUtils/moveDirectory (File. tmproot) (File. stormroot))
          (do
            (FileUtils/forceMkdir (File. stormroot))
            (Files/move (.toPath (File. tmproot)) (.toPath (File. stormroot))
                        (doto (make-array StandardCopyOption 1) (aset 0 StandardCopyOption/ATOMIC_MOVE)))))
```

So I think the resulting code should be more like

```
            if (Utils.isOnWindows()) {
              // Files/move with non-empty directory doesn't work well on Windows
              FileUtils.moveDirectory(new File(tmproot), new File(stormroot));
            } else {
              FileUtils.forceMkdir(new File(stormroot));
              Files.move(new File(tmproot).toPath(), new File(stormroot).toPath(), StandardCopyOption.ATOMIC_MOVE);
            }
```